Studio-Lovelies/GG-JointJustice-Unity
# Re-introduce automated testing

## Summary
With this PR, EditMode and PlayMode tests are ran in parallel to each build via GitHub actions. [[successful example](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/runs/4865809510?check_suite_focus=true)] [[failed example](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/runs/4865659605?check_suite_focus=true)]
(Additionally, code coverage reports for the tests is stored as build artifacts and also posted to [codecov.io](https://app.codecov.io/gh/Studio-Lovelies/GG-JointJustice-Unity). [[example](https://app.codecov.io/gh/ViMaSter/GG-JointJustice-Unity/commits?page=1)])

This change also splits a subset of the `AppearingDialogueController` test suite into two separate categories:
- [An isolated set of tests for verifying `GetDelay()` returns the correct delay for a given character based on the currently set flags](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/autotest/unity-ggjj/Assets/Tests/PlayModeTests/Scripts/AppearingDialogueController/GetDelay.cs)
- [One integration test which asserts that if `GetDelay()` is bigger, it takes longer for the text box to render content](https://github.com/Studio-Lovelies/GG-JointJustice-Unity/blob/autotest/unity-ggjj/Assets/Tests/PlayModeTests/Scripts/AppearingDialogueController/IntegrationTests.cs#L35-L72)

This change appears to be needed for tests on GitHub Actions to reflect the stability of running them locally. 

## Before/after screenshots and/or animated gif
### Before
![image](https://user-images.githubusercontent.com/1689033/150485836-2a1c6d38-f55b-4c86-9dfa-3efb8cc8f43b.png)

### After
![image](https://user-images.githubusercontent.com/1689033/150485859-5f28fd1e-0f40-4b0d-b224-1c60061a6d81.png)
![image](https://user-images.githubusercontent.com/1689033/150485880-879bd220-6c02-4825-af27-79ec2146a319.png)

## Testing instructions
1. Fork this branch of the repository to a local clone
2. Append an empty commit to this branch on your local clone
3. Notice automated tests being executed in parallel to the game build and displaying 

## Additional information

- `[ ]` Changes UI
- `[x]` Introduces new feature
  - Automated tests inside GitHub Actions builds
  - Code Coverage reports at [codecov.io](https://app.codecov.io/gh/Studio-Lovelies/GG-JointJustice-Unity)
- `[ ]` Removes existing feature
- `[ ]` Has associated resource:
